### ISSUE TITLE
Improve error message for feature flag incompatibility

### ISSUE THREAD

**Issue by hackaugusto** [0xPolygonMiden/miden-node#254]
Either the client or the server may use the `testing` flag, which changes some settings in the system to reduce PoW and allow for faster testing.

The issue is that when mixing the two systems, the error message is pretty opaque:

```
Feb 29 14:31:22 miden-node-1 miden-node[102235]: 2024-02-29T14:31:22.546347Z ERROR rpc:submit_proven_transaction: miden-rpc: rpc/src/server/api.rs:116: error: status: InvalidArgument, message: "Invalid transaction proof for transaction: 0xeb9a5472ac63beff9c98dabed48690aca7f8e7bed3306e963c30295ad21cf2a0", details: [], metadata: MetadataMap { headers: {} }
```

If possible we should add more detail to the message, to help debug issues.

**Comment by hackaugusto** [comment]
I guess a simple hack would be to try verifying the proof with both settings, the issue with that approach is that this behavior is undesirable in `prod`. @bobbinth do we have something to extract details from the verifier? Something to narrow down the problem space would be useful.

**Comment by bobbinth** [comment]
I wonder if this is somewhat related to #255. Basically, if we send something like "protocol version" with every request, we'd be able to tell whether the client and the node are incompatible.

**Comment by hackaugusto** [comment]
Yeah, versioning the protocol would also fix the issue here. We have to decide on having something more connection oriented or something more datagram oriented. By that I mean either a handshake with version negotiation as the first messages in a connection, or having a packet header for every packet with their own version. Both approaches have their pros/cons.

- handshake
   - pros: server/client agree on capabilities upfront. Making it easier to abort if something is missing. For example, if the server doesn't have an endpoint the client expects.
   - cons: adds complexity to the connection management, down the road may require server pinning
- datagram:
   - pros: this fits a graceful feature degradation more easily, also gives the ability to version the messages basically for free
   - cons: this will have a small cost in size (I think it is negligible, but pointing it out). if we don't take care, the protocol can become too complicated, if there are too many versions

Another approach is to make the settings explicit in the protocol, instead of implicit. By that I mean we can add metadata to the message, which describes the proof settings used by the client. The error messages would be the best in this scenario.

**Comment by hackaugusto** [comment]
Perhaps, for debugging purposes, the best approach is the metadata. We can make that optional in protobuf so that in the future we don't pay the space cost. And we can enable it while we are developing, to get better error messages.

**Comment by bobbinth** [comment]
Conceptually, i like the idea of a handshake a bit better. Not only can we communicate protocol version (as well as network type - e.g., devnet, testnet etc.), but we could also extend it in the future to include more info regarding capabilities of the node (e.g., there could be archive nodes, full nodes, nodes that have pruned note data to the last year or w/e else). And sending all this data with every requests seems like an overkill.

But on the other hand, I'm not sure how to make it work with how our client works currently (or at least the CLI version). There, we don't hold a persistent connection - so, not exactly clear when the handshake should take place.

**Comment by hackaugusto** [comment]
> But on the other hand, I'm not sure how to make it work with how our client works currently (or at least the CLI version). There, we don't hold a persistent connection - so, not exactly clear when the handshake should take place.

Well, in that case every use of the CLI would need a handshake.

**Comment by bobbinth** [comment]
> Well, in that case every use of the CLI would need a handshake.

Maybe that's fine because the issue is really specific to our client CLI implementation. We can imagine another implementation (e.g., the CLI keeps a persistent connection) where this won't be an issue.

**Comment by Mirko-von-Leipzig** [comment]
Closing since the `testing` feature is no longer around (#619) :)

There may still be relevant content here re: handshaking and protocol version verification, but we should create a new issue for that.
